Integrate navigator observer

Guardiola31337 opened the pull request:
### Description
Supersedes https://github.com/mapbox/mapbox-navigation-android/pull/4354

This builds on top of https://github.com/mapbox/mapbox-navigation-android/pull/4417

- Now `NavigatorObserver` [`onStatus`](https://github.com/mapbox/mapbox-navigation-native/blob/cf2fcee5bf9b768b68eb49919711bd28b06d1b71/bindings/generated/cpp/mapbox/navigation/navigator_observer.hpp#L22) callback returns `NavigationStatusOrigin` in addition to `NavigationStatus`. This is an `enum` including information about where this status is coming from (`LOCATION_UPDATE`, `LEG_CHANGE`, `SET_ROUTE` and `UNCONDITIONAL`). Is that correct @DmitryAk? How can we use this extra information for our benefit from the SDK side? Should we expose it to developers? Could you talk more about why you added this in the first place? Currently, we're just ignoring this value.

**TO-DO**

- [x] Expose  `NavigatorConfig` `PollingConfig` through `NavigationOptions`?
- [x] Fix `MapboxTripSession` `@Ignore`d tests
- [x] Extensive field testing
- [x] @DmitryAk Any other things that we need to think about? Please, feel free to add any checks here or as a comment.

### Changelog
<!--
Include changelog entry (e.g. Fixed an unexpected change in recenter button when resuming the app.).
See https://github.com/mapbox/navigation-sdks/blob/main/documentation/android-changelog-guidelines.md.
You can remove the changelog block and add a `skip changelog` label, when applicable.
 -->
```
<changelog>Migrated to callback-based native getStatus approach.</changelog>
<changelog>Removed `routeGeometryWithBuffer` from `RouteProgress`.</changelog>
<changelog>:warning: Internal `setUnconditionalPollingPatience` and `setUnconditionalPollingInterval` have been moved to `InternalUtils` `object` and have to be called _before_ `MapboxNavigation` is instantiated to actually take effect.</changelog>
```

cc @etl @mskurydin @SiarheiFedartsou @truburt @zugaldia

DmitryAk reviewed (commented):
LGTM from NavNative point of view

RingerJK commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:
```diff
@@ -183,11 +161,42 @@ internal class MapboxTripSession(
+    private val navigatorObserver = object : NavigatorObserver() {
+        override fun onStatus(origin: NavigationStatusOrigin, status: NavigationStatus) {
+            mainJobController.scope.launch {
+                val tripStatus = MapboxNativeNavigatorImpl.generateTripStatusFrom(status)
```
why don't check `!isActive`? 🤔

Guardiola31337 commented:
Addressed all feedback from https://github.com/mapbox/mapbox-navigation-android/pull/4354

@kmadsen @LukasPaczos @RingerJK This is ready for another round of 👀

Guardiola31337 commented:
cc @mapbox/navigation-ios

LukasPaczos commented on libnavigator/src/main/java/com/mapbox/navigation/navigator/internal/MapboxNativeNavigator.kt:
```diff
@@ -76,21 +75,9 @@ interface MapboxNativeNavigator {
-     * and verify that the user is still on the route. This method also determines
-     * if an instruction needs to be called out for the user.
-     *
-     * @param navigatorPredictionMillis millis for navigation status predictions.
```
Same as above, the prediction value should now be passed to `PollingConfig`.

LukasPaczos replied:
This is important to address so that we can effectively support use-cases with higher frequency signal than 1Hz.

Guardiola31337 replied:
Added `InternalUtils` https://github.com/mapbox/mapbox-navigation-android/blob/b2ad867af1d4f2dd4851618c354b7ebcaa3a3433/libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/utils/InternalUtils.kt Please let me know what you think 🙏

Guardiola31337 replied:
cc @DmitryAk

DmitryAk replied:
@Guardiola31337 @LukasPaczos please see my comment https://github.com/mapbox/mapbox-navigation-android/pull/4419/files#r655903753
It looks like previously `InternalUtils` methods accepted values in milliseconds, now they accept values in seconds. It's important to not miss this, so the `InternalUtils` consumers behavior will not be broken.

LukasPaczos reviewed (commented):
I found one small issue where we miss a status update:
1. Start trip session.
2. Get location updates.
3. Set a route (new banner instructions are available).
4. Stop trip session.
5. Change a route.
6. Start trip session but do not provide any location updates.

This will leave us with an old route progress object (and old banner instructions) because when step 4 happened and we changed the route, the `NavigatorObserver` was unsubscribed. We need to wait for the next location update to come (or `lastLocation` update) to receive the new banner instructions. I don't think this is necessarily a blocker, we clear all states when stopping a session, so maybe it's expected that we need to wait for another location update? Just flagging.

@DmitryAk I see this comment in the `NavigatorObserver#onStatus`:
```
NB: There is no guarantee about thread where it can be called. So, synchronization is up to SDK.
```
Is this still true or is it outdated?

Also @DmitryAk are there any additional conditions for the automatic polling to kick in? I'm not getting any status updates after pausing the raw location signal.

DmitryAk commented:
> @DmitryAk I see this comment in the `NavigatorObserver#onStatus`:
> 
> ```
> NB: There is no guarantee about thread where it can be called. So, synchronization is up to SDK.
> ```
> 
> Is this still true or is it outdated?
> 
> Also @DmitryAk are there any additional conditions for the automatic polling to kick in? I'm not getting any status updates after pausing the raw location signal.

@LukasPaczos 
1. The comment is outdated, I've just updated it - the updated comment should be available with the next NN release.
2. unconditional (automatic) polling is controlled with `unconditionalStatusPollingPatienceSeconds` and `unconditionalStatusPollingIntervalSeconds` settings in config. If it is not observed with proper configuration - looks like a bug, need to investigate it.

Guardiola31337 commented:
@LukasPaczos @DmitryAk 
> are there any additional conditions for the automatic polling to kick in? I'm not getting any status updates after pausing the raw location signal.

Are you referring to prediction? As far as I know prediction is currently disabled altogether. How does that relate to the unconditional polling options?

Cc @mskurydin @SiarheiFedartsou @etl

LukasPaczos commented:
> Are you referring to prediction? As far as I know prediction is currently disabled altogether. How does that relate to the unconditional polling options?

Are predictions really disabled @DmitryAk? Why is that if there's an option for it?

I wasn't talking about prediction time though, but the unconditional polling that wasn't firing for me after location updates were paused for more than 2s (which is documented as default).

SiarheiFedartsou commented:
> > Are you referring to prediction? As far as I know prediction is currently disabled altogether. How does that relate to the unconditional polling options?
> 
> Are predictions really disabled @DmitryAk? Why is that if there's an option for it?
> 
> I wasn't talking about prediction time though, but the unconditional polling that wasn't firing for me after location updates were paused for more than 2s (which is documented as default).

No, prediction is not disabled. Prediction can be controlled via lookahead field in PollingConfig:
https://github.com/mapbox/mapbox-navigation-native/blob/10c3e85946ff41d9ebc93b4be1d5584bb3fa69bc/bindings/navigator.mbg#L356

DmitryAk commented:
> Are predictions really disabled @DmitryAk? Why is that if there's an option for it?
> 
> I wasn't talking about prediction time though, but the unconditional polling that wasn't firing for me after location updates were paused for more than 2s (which is documented as default).

As @SiarheiFedartsou mentioned, prediction is not disabled, so "unconditional" `onStatus` updates should work. 
(just to clarify: raw location "extrapolation" is not required anymore - maybe, this was misunderstood as "prediction", but it is not "prediction". Previously if we received `UpdateLocation`, and you called `GetStatus` 0.5 sec later - the raw location was "extrapolated" 0.5 seconds forward - but this is no longer required due to `GetStatus` is calculated immediately on the raw location)

@LukasPaczos maybe, 2s of waiting for them was not enough? Could you please clarify: is this reproducible for longer periods of waiting? BTW: on MacOS there was an issue with callbacks scheduling delaying - callbacks were executed with significant delay (up to 4 seconds) - maybe, you just observed the same. This issue was fixed, will be available with the next release. But maybe this is not the case for Android. Another possible reason: was the `main` looper available for the callbacks executing?

Guardiola31337 commented:
Below you can see a video going through a tunnel and the associated history file 👀 

https://user-images.githubusercontent.com/1668582/119695309-09572580-be1c-11eb-9870-729a85e3b961.mov

[tunnel.json.zip](https://github.com/mapbox/mapbox-navigation-android/files/6548222/tunnel.json.zip)

This is forcing `updateLocation`s to be every 3 seconds. As seen in the video (location update > puck stuck for 3 seconds in the last location > jump when next location update received) `lookAhead` doesn't work neither in normal roads nor in tunnels. Worth mentioning that I tried to reproduce https://github.com/mapbox/mapbox-navigation-android/pull/4419#pullrequestreview-666829435 and I'm always receiving unconditional updates through `NavigatorObserver#onStatus`, although based on the video, of the same location fix.

How do `lookAhead` and the unconditional updates work exactly? Could you clarify this further? Not sure if I completely follow https://github.com/mapbox/mapbox-navigation-android/pull/4419#issuecomment-848751150 and if the behavior we're seeing is expected.

cc @mskurydin @SiarheiFedartsou @DmitryAk @LukasPaczos

Guardiola31337 commented:
This is the `diff` I used for testing https://github.com/mapbox/mapbox-navigation-android/pull/4419#issuecomment-848921641

```diff
diff --git a/examples/src/main/java/com/mapbox/navigation/examples/core/camera/MapboxCameraAnimationsActivity.kt b/examples/src/main/java/com/mapbox/navigation/examples/core/camera/MapboxCameraAnimationsActivity.kt
index 00bbe179e..8789d6f7e 100644
--- a/examples/src/main/java/com/mapbox/navigation/examples/core/camera/MapboxCameraAnimationsActivity.kt
+++ b/examples/src/main/java/com/mapbox/navigation/examples/core/camera/MapboxCameraAnimationsActivity.kt
@@ -382,6 +382,7 @@ class MapboxCameraAnimationsActivity :
                 .locationEngine(ReplayLocationEngine(mapboxReplayer))
                 .build()
         ).apply {
+            toggleHistory(true)
             registerLocationObserver(
                 object : LocationObserver {
 
@@ -606,6 +607,7 @@ class MapboxCameraAnimationsActivity :
     override fun onDestroy() {
         super.onDestroy()
         binding.mapView.onDestroy()
+        val history = mapboxNavigation.retrieveHistory()
         mapboxNavigation.onDestroy()
     }
 
diff --git a/libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt b/libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt
index e78b7f4a8..1d644bd5a 100644
--- a/libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt
+++ b/libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt
@@ -3,6 +3,7 @@ package com.mapbox.navigation.core.trip.session
 import android.hardware.SensorEvent
 import android.location.Location
 import android.os.Looper
+import android.util.Log
 import com.mapbox.android.core.location.LocationEngineCallback
 import com.mapbox.android.core.location.LocationEngineResult
 import com.mapbox.api.directions.v5.models.BannerInstructions
@@ -169,6 +170,7 @@ internal class MapboxTripSession(
 
     private val navigatorObserver = object : NavigatorObserver() {
         override fun onStatus(origin: NavigationStatusOrigin, status: NavigationStatus) {
+            Log.e("DEBUG", "DEBUG onStatus=$status")
             val tripStatus = status.getTripStatusFrom(route)
             val enhancedLocation = tripStatus.navigationStatus.location.toLocation()
             val keyPoints = tripStatus.navigationStatus.keyPoints.toLocations()
@@ -478,13 +480,20 @@ internal class MapboxTripSession(
         }
     }
 
+    private var currentTimeMillis = System.currentTimeMillis()
+
     private fun updateRawLocation(rawLocation: Location) {
         if (state != TripSessionState.STARTED) return
 
         this.rawLocation = rawLocation
         locationObservers.forEach { it.onRawLocationChanged(rawLocation) }
         mainJobController.scope.launch {
-            navigator.updateLocation(rawLocation.toFixLocation())
+            val current = System.currentTimeMillis()
+            if (current - currentTimeMillis > 3000) {
+                currentTimeMillis = current
+                Log.e("DEBUG", "DEBUG rawLocation=$rawLocation")
+                navigator.updateLocation(rawLocation.toFixLocation())
+            }
         }
     }
```

DmitryAk commented:
> This is forcing `updateLocation`s to be every 3 seconds. As seen in the video (location update > puck stuck for 3 seconds in the last location > jump when next location update received) `lookAhead` doesn't work neither in normal roads nor in tunnels. Worth mentioning that I tried to reproduce [#4419 (review)](https://github.com/mapbox/mapbox-navigation-android/pull/4419#pullrequestreview-666829435) and I'm always receiving unconditional updates through `NavigatorObserver#onStatus`, although based on the video, of the same location fix.

what I can see from the attached `tunnel.json.zip` - is that per each `updateLocation` there are 3 `getStatus` events with different locations - so it looks like "unconditional onStatus notification" and "lookAhead" both working.
Also it is visible from the traces that sometimes our simulation goes "too far" and the next "update location" is closer to the previous "update location" than the last "getStatus" - this is not good. But I don't see that "lookAhead doesn't work". You could see that there are different "getStatus" locations per each "updateLocation" below.
Need to investigate this case further (next week).
![Screen Shot 2021-05-28 at 22 30 02](https://user-images.githubusercontent.com/10007450/120033517-539afb00-c004-11eb-9d70-ed5cc0e4aa88.png)

LukasPaczos commented:
@DmitryAk here's an example where unconditional polling doesn't kick in:
[history_unconditional.zip](https://github.com/mapbox/mapbox-navigation-android/files/6569767/history_unconditional.zip)
![Screenshot from 2021-05-31 12-52-52](https://user-images.githubusercontent.com/16925074/120182789-23737800-c20f-11eb-8c1c-63459e17c3ae.png)

LukasPaczos commented:
Or to say it differently - the status is updated, but it's constant and does not simulate. What are the criteria for simulation in this case?

LukasPaczos commented:
My diff for reference:
```diff
diff --git a/libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt b/libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt
index e78b7f4a8..3f900ba26 100644
--- a/libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt
+++ b/libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt
@@ -478,6 +478,7 @@ internal class MapboxTripSession(
         }
     }
 
+    var counter = 0
     private fun updateRawLocation(rawLocation: Location) {
         if (state != TripSessionState.STARTED) return
 
@@ -485,6 +486,12 @@ internal class MapboxTripSession(
         locationObservers.forEach { it.onRawLocationChanged(rawLocation) }
         mainJobController.scope.launch {
             navigator.updateLocation(rawLocation.toFixLocation())
+            if (route != null) {
+                counter++
+            }
+            if (counter >= 10) {
+                stopLocationUpdates()
+            }
         }
     }
```

DmitryAk commented:
> Or to say it differently - the status is updated, but it's constant and does not simulate. What are the criteria for simulation in this case?

@LukasPaczos @Guardiola31337 currently we have a limit for simulation max time (7 seconds by default): https://github.com/mapbox/mapbox-navigation-native/blob/512071b38e25ce38c1572fab70c15c163ed61a02/src/ride/config.hpp#L132 - this was done few months ago to avoid strange issues (when moving in tunnel is simulated, but actually the user is not moving, or moving to another exit, etc.)
Looks like on your last trace simulation worked for 7 seconds, and then - stopped. This corresponds to current design.

LukasPaczos commented:
> Looks like on your last trace simulation worked for 7 seconds

The status updates were provided for 7 seconds, but they didn't simulate, all the updates were the same. Is that expected?

DmitryAk commented:
> > Looks like on your last trace simulation worked for 7 seconds
> 
> The status updates were provided for 7 seconds, but they didn't simulate, all the updates were the same. Is that expected?

@LukasPaczos 
This is what I got when replaying the trace in qt_player with default config:
![Screen Shot 2021-05-31 at 16 05 25](https://user-images.githubusercontent.com/10007450/120197873-2def3b00-c22a-11eb-9120-67c80483ca85.png)
^ the selected event (big red circle) is the last `UpdateLocation` event. After it there are 7 different simulated `GetStatus` events (small blue points on the line), and on the last simulated `GetStatus` there are 15 more `GetStatus` events, but they are pointing to the same location - i.e., simulation was stopped, but still there were unconditional `GetStatus` updates. This corresponds to the expected behavior.
I see that your actual trace doesn't contain the default sumultaion

![Screen Shot 2021-05-31 at 16 20 56](https://user-images.githubusercontent.com/10007450/120199748-41030a80-c22c-11eb-8517-40bbcbcdb1dd.png)
it looks similar to when the https://github.com/mapbox/mapbox-navigation-native/blob/512071b38e25ce38c1572fab70c15c163ed61a02/src/ride/config.hpp#L132 was set to 0. Could it be the case?

LukasPaczos commented:
Thanks @DmitryAk, looks like an issue on our side, we pass `false` for `noSignalSimulationEnabled`:
https://github.com/mapbox/mapbox-navigation-android/blob/a29fc2e301880748ab67b9d1a90c61532dce975d/libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt#L204.

Could you fix it up as part of this PR @Guardiola31337?

Guardiola31337 commented:
> looks like an issue on our side, we pass `false` for `noSignalSimulationEnabled`:
https://github.com/mapbox/mapbox-navigation-android/blob/a29fc2e301880748ab67b9d1a90c61532dce975d/libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt#L204.
>
> Could you fix it up as part of this PR @Guardiola31337?

```java
    /**
     * if `true` NN will try to simulate puck movement in case of no signal
     * default: `false` for automotive profile, `true` for mobile profile
     */
```

We should probably pass `null` and let NN to handle it as we're currently configuring the profile properly 👀 https://github.com/mapbox/mapbox-navigation-android/blob/a29fc2e301880748ab67b9d1a90c61532dce975d/libnavigator/src/main/java/com/mapbox/navigation/navigator/internal/NavigatorLoader.kt#L61-L71

What do you think @DmitryAk?

cc @mskurydin @SiarheiFedartsou

Guardiola31337 commented:
https://github.com/mapbox/mapbox-navigation-android/pull/4419#issuecomment-851887726 BTW this is what we're doing with the rest of the options that we're not exposing.

Guardiola31337 commented:
Tested https://github.com/mapbox/mapbox-navigation-android/pull/4419#issuecomment-848928820 again passing `null` for `noSignalSimulationEnabled` and this is what we got 👀 

https://user-images.githubusercontent.com/1668582/120288598-645ea180-c28e-11eb-8ce6-c3fd166bfb59.mp4

Noting that this also goes through a tunnel.

Apparently it seems to work although there's still some (expected?) "jumpiness". Attaching the history file for reference 

[tunnel_madrid.json.zip](https://github.com/mapbox/mapbox-navigation-android/files/6574549/tunnel_madrid.json.zip)

Going to test https://github.com/mapbox/mapbox-navigation-android/pull/4419#issuecomment-851448589

Guardiola31337 commented:
> Going to test https://github.com/mapbox/mapbox-navigation-android/pull/4419#issuecomment-851448589

<img width="1645" alt="Screen Shot 2021-06-01 at 10 29 00 AM" src="https://user-images.githubusercontent.com/1668582/120292152-0d5acb80-c292-11eb-8062-dc7fdc61ad81.png">

@DmitryAk @LukasPaczos how does that look to you?

[tunnel_madrid_stop_after_15_seconds.json.zip](https://github.com/mapbox/mapbox-navigation-android/files/6574674/tunnel_madrid_stop_after_15_seconds.json.zip)

History file for reference ☝️

DmitryAk commented:
> @DmitryAk @LukasPaczos how does that look to you?

it looks to me as expected, looks like simulation worked here, LGTM

LukasPaczos commented:
The routes alternatives test seems to consistently fail on this branch. It fails for me locally as well, not sure if related to this PR though.

LukasPaczos commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/utils/InternalUtils.kt:
```diff
@@ -1,27 +0,0 @@
- *
- * Do not use in a production environment.
- */
-fun setUnconditionalPollingPatience(patience: Long?) {
-    MapboxTripSession.UNCONDITIONAL_STATUS_POLLING_PATIENCE = patience ?: 2000L
-}
-
-/**
- * Internal API used for testing. Sets the static unconditional polling interval value for all
- * instances in the process.
- *
- * Pass `null` to reset to default.
- *
- * Do not use in a production environment.
- */
-fun setUnconditionalPollingInterval(interval: Long?) {
-    MapboxTripSession.UNCONDITIONAL_STATUS_POLLING_INTERVAL = interval ?: 1000L
-}
```
If we don't want to expose this publicly, we should keep this internal utility anyway and set the value to the navigator config. Some customers are relying on this for testing purposes.

Guardiola31337 replied:
Added `InternalUtils` https://github.com/mapbox/mapbox-navigation-android/blob/b2ad867af1d4f2dd4851618c354b7ebcaa3a3433/libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/utils/InternalUtils.kt Please let me know what you think 🙏

Guardiola31337 replied:
cc @DmitryAk

Guardiola31337 commented:
> The routes alternatives test seems to consistently fail on this branch. It fails for me locally as well, not sure if related to this PR though.

Yeah, I also noticed https://github.com/mapbox/mapbox-navigation-android/issues/4429

LukasPaczos commented:
It looks like it might be a regression. It'd be great to get to the bottom of this before merging.

kmadsen commented on instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt:
```diff
@@ -59,11 +59,15 @@ class RouteAlternativesTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::cla
         val routes = requestDirectionsRouteSync(coordinates).reversed()
 
+        // Warm up the locations
+        runOnMainSync {
```
this block is not needed. but going to let the current test run complete first

Guardiola31337 commented:
> It looks like it might be a regression. It'd be great to get to the bottom of this before merging.

As mentioned in https://github.com/mapbox/mapbox-navigation-android/issues/4429#issuecomment-865133144 we should refactor these tests (likely we've been lucky all the time with those tests but the tests are not "correct" and with callback approach it's more evident) but that can happen at a later time and shouldn't block this PR from merging.

Ready for a final round of 👀 

cc @kmadsen @DmitryAk @LukasPaczos

DmitryAk commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/utils/InternalUtils.kt:
```diff
@@ -1,27 +1,49 @@
+     * Do not use in a production environment.
+     */
+    fun setUnconditionalPollingPatience(patienceSeconds: Double?) {
+        UNCONDITIONAL_POLLING_PATIENCE_SECONDS = patienceSeconds ?: 2.0
```
@Guardiola31337 It looks like previously this method accepted `patience: Long?` milliseconds value, and now it accepts `patienceSeconds: Double?` seconds. So if the consumers of this interface are not notified about this change - they could get broken behavior (e.g., they could just continue to call `setUnconditionalPollingPatience(1000)` - this will be compiled, but this will break the old behavior, because now it means 1000 seconds, not 1000 milliseconds). Suggestions:
1. Don't update this interface, keep old `setUnconditionalPollingPatience(patience: Long?)` signature. Convert the incoming `milliseconds` value into `seconds` inside the method.
2. Add documentation about `seconds` or `milliseconds` value usage - so this is declared explicitly.

LukasPaczos replied:
I'd be behind leaving the milliseconds in the public API and converting the value internally into seconds.

Guardiola31337 replied:
@DmitryAk Going to refactor in favor of

> 1. Don't update this interface, keep old `setUnconditionalPollingPatience(patience: Long?)` signature. Convert the incoming `milliseconds` value into `seconds` inside the method.

In any case, why did that change internally? I'd expect that NN also accepts milliseconds so we don't need to do any conversion from the SDK side ¯\\\_(ツ)\_/¯

Guardiola31337 replied:
@LukasPaczos @DmitryAk I've just pushed a commit addressing this. Ready for another round of 👀

DmitryAk replied:
@Guardiola31337 I cannot find: how were these values accepted by NN previously (in milliseconds)? The only previous usage I found is `delay(UNCONDITIONAL_STATUS_POLLING_PATIENCE)`, but it is internal to Android - so it looks like these values were not accepted by NN previously.

Guardiola31337 replied:
Yeah but now they're requested from `PollingConfig`. Should these be in milliseconds too? Currently they're in seconds.

SiarheiFedartsou replied:
I'd vote for nanoseconds - it is used for location timestamp, feels more natural to make it nanoseconds too then

Guardiola31337 replied:
Happy to change in the next release when that fix is available from NN 👍

kmadsen commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/utils/InternalUtils.kt:
```diff
@@ -1,27 +1,49 @@
 package com.mapbox.navigation.core.internal.utils
 
-import com.mapbox.navigation.core.trip.session.MapboxTripSession
+object InternalUtils {
```
I don't think we should introduce a new way to set options here. This pattern is up for discussion because it has individually mutable parameters. Also it's unclear why they're `@Volatile`, it's as if they are built for thread safety when the initialization has strict requirements - they need to be initialized when `MapboxNavigation` is constructed or reconstructed.

I think we should simply define a constant

``` kotlin
val nativePollingOptions = PollingConfig(
    LOOK_AHEAD_SECONDS,
    POLLING_PATIENCE_SECONDS,
    POLLING_INTERVAL_SECONDS
)
```

If we want to make these configurable, we should follow the consistent pattern and then make an Options class. If we want that to be internal, change the visibility to `internal`

``` kotlin
internal class StatusPollingOptions private constructor(
  val Double lookAhead,
  val Double unconditionalPatience,
  val Double: unconditionalInterval
) {
  Builder {
    private var Double lookAhead = 1.0
    private var Double unconditionalPatience = 2.0
    private var Double: unconditionalInterval = 2.0

    fun toBuilder(): Builder = Builder().apply { ...

    fun lookAhead(lookAhead: Double) = apply {
        this.lookAhead = lookAhead
    }
    fun unconditionalPatience(unconditionalPatience: Double) = apply {
        this.unconditionalPatience = unconditionalPatience
    }
    fun unconditionalInterval(unconditionalPatience: Double) = apply {
        this.unconditionalInterval = unconditionalInterval 
    }

    fun build(): StatusPollingOptions
...

    private val pollingConfig = PollingConfig(
        statusPollingOptions.lookAhead,
        statusPollingOptions.unconditionalPatience,
        statusPollingOptions.unconditionalInterval
    )
```

if we want to include a level of abstraction from nav native (e.g., milliseconds instead of seconds) Add a mapper. That way we're not adding more logic to `MapboxNavigation` that is untested

``` kotlin
object PollingConfigMapper {
    fun mapFrom(statusPollingOptions: StatusPollingOptions) = PollingConfig(
        lookAhead = statusPollingOptions.lookAheadMillis / ONE_SECOND_IN_MILLIS,
        ...
    )
}
```

kmadsen replied:
Ah and see, the old way was making them `@Volatile` because they were watched by different threads
https://github.com/mapbox/mapbox-navigation-android/pull/4419/files#diff-37365d0e69a56e8efddf9ff8e735c04b3474c7df666259edfbc961855c3556a2L543

The new way requires the fields to be created during construction time, so we no longer need them to be `@Volative` because we no longer can dynamically change the polling rates

Guardiola31337 replied:
Removed `@Volatile`.

LukasPaczos commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/utils/InternalUtils.kt:
```diff
@@ -1,27 +1,44 @@
+    internal var UNCONDITIONAL_POLLING_PATIENCE_MILLISECONDS = 2000L
+    internal var UNCONDITIONAL_POLLING_INTERVAL_MILLISECONDS = 1000L
+
+    /**
+     * Internal API used for testing. Sets the static look ahead value for all
+     * instances in the process. Needs to be set prior to [MapboxNavigation] creation.
+     *
+     * Pass `null` to reset to default.
+     *
+     * Do not use in a production environment.
+     */
+    fun setLookAhead(lookAheadInMillis: Long?) {
+        LOOK_AHEAD_MILLISECONDS = lookAheadInMillis ?: 1000L
+    }
```
This parameter is important in scenarios where location is coming at a higher frequency than 1Hz. We should make it a part of the public API as it was before.

LukasPaczos commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:
```diff
@@ -529,71 +531,15 @@ internal class MapboxTripSession(
-    }
-
-    private suspend fun getNavigatorStatus(): TripStatus {
-        return navigator.getStatus(navigationOptions.navigatorPredictionMillis)
```
As mentioned above, the `navigationOptions.navigatorPredictionMillis` should still be respected and public.

LukasPaczos commented:
Let's add to the changelog that the internal `setUnconditionalPollingPatience` and `setUnconditionalPollingInterval` have been moved to `InternalUtils` `object` and have to be called _before_ `MapboxNavigation` is instantiated to actually take effect. Even though it's an internal API, some customers are using it in test setups.
